# Multiclass Classification with SVM
The application of SVM classifier in datasets that have one or more classes is possible. Conventional approaches that many people tend to take include one-VS-rest and one-VS-one. In One-VS-rest, the goal is to create K SVM binary classifiers. The Kth classifier is the positive case and the rest of the classes are considered negative.
One-vs-one strategy utilized pairwise comparison by constructing a set of SVM classifiers that can differentiation data points from each pair of classes.

## Wine Dataset 

In [2]:
from sklearn.datasets import load_wine

wine_data = load_wine()

X = wine_data.data
y = wine_data.target

# Data Shape
print('Shape of X: ', X.shape)
print('Shape of y: ', y.shape)

# Label Names
print('Target Labels: ', wine_data.target_names)

# Negative Classes
class0 = (y == 0).sum()
class1 = (y == 1).sum()
class2 = (y == 2).sum()

print(f'{class0} class0 samples,\n{class1} class1 samples,\n{class2} class2 samples.')

Shape of X:  (178, 13)
Shape of y:  (178,)
Target Labels:  ['class_0' 'class_1' 'class_2']
59 class0 samples,
71 class1 samples,
48 class2 samples.


In [3]:
# Training and Testing Data
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

# Initialize Classifier and Fit Model
model = SVC(kernel='linear', C=1.0, random_state=42)
model.fit(X_train, y_train)

#Check Model accuracy

model_accuracy = model.score(X_test, y_test)
print(f'Accuracy: {model_accuracy*100:1f}%')

Accuracy: 97.777778%


The model accuracy is 97.8%, which confirms that SVM works even in multiclass classification. Classification report can bring into light how the model performs for the individual classes.

In [4]:
from sklearn.metrics import classification_report

predict = model.predict(X_test)

print(classification_report(y_test, predict))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        15
           1       1.00      0.94      0.97        18
           2       0.92      1.00      0.96        12

    accuracy                           0.98        45
   macro avg       0.97      0.98      0.98        45
weighted avg       0.98      0.98      0.98        45

